PostCSS style transformation: fail gracefully instead of throwing an error

zaguiini opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/56082.

## What?

https://github.com/WordPress/gutenberg/pull/49521 introduces a regression in which invalid CSS throws an error. This PR fixes the problem by letting the transformation fail gracefully.

## Why?

https://github.com/WordPress/gutenberg/pull/49521 changed the internal CSS transformer to fix a couple of issues, but invalid CSS handling was not exhaustively tested.

## How?

I'm wrapping the transformation function with a try-catch block and returning null in case it fails to parse the CSS.

## Testing Instructions

Follow the instructions in https://github.com/WordPress/gutenberg/issues/56082.

zaguiini committed:
PostCSS style transformation: fail gracefully instead of throwing error

zaguiini commented:
@strarsis I'd appreciate a review here.

strarsis commented:
@zaguiini: Great addition! :thumbsup:

afercia commented on packages/block-editor/src/utils/transform-styles/index.js:
```diff
@@ -19,18 +19,30 @@ import rebaseUrl from 'postcss-urlrebase';
+				].filter( Boolean )
+			).process( css, {} ).css; // use sync PostCSS API
+		} catch ( error ) {
+			if ( ! ( error instanceof CssSyntaxError ) ) {
```
I'm not familiar enough with postCSS to know when an error is _not_ a CssSyntaxError. In what scenario this condition will be true and the console warn be useful?

afercia commented on packages/block-editor/src/utils/test/transform-styles.js:
```diff
@@ -4,6 +4,28 @@
 import transformStyles from '../transform-styles';
 
 describe( 'transformStyles', () => {
+	it( 'should not throw error in case of invalid css', () => {
```
This test is welcome but it actually doesn't not test what was broken in the UI of the custom CSS advanced panel. It just tests the return value is `null` when there is a CSS error. It doesn't make sure the UI works as expected.

afercia commented:
Quickly tested and it does fix the issue.

I'd recommend to add an E2E test for the feature that was actually broken in `packages/block-editor/src/components/global-styles/advanced-panel.js`. The test should make sure that:
- When there is a CSS error and the CSS textarea is blurred, the icon notice appears.
- When the CSS error is fixed and the CSS textarea is blurred, the icon notice disappears.

afercia commented:
On a related note: Previously, when there was an error in the CSS added to the textarea, the console warn gave some hint on the type of error:

![Screenshot 2023-11-15 at 09 49 11](https://github.com/WordPress/gutenberg/assets/1682452/028e7471-ba68-4396-84f8-82ceac21ac8e)

We could argue that wasn't that useful to average users. However, now there won't be anything in the console. I read there are nice ways to handle the error output, see https://postcss.org/docs/postcss-runner-guidelines#output
As a follow-up, I'd suggest to try to add some more meaningful error message in the UI, or at least in the console. I quickly tried with this:

Input:
```
body {background: red;}
body {background: aqua
```

Dupming this to the console:
`console.log( error.message + '\n' + error.showSourceCode() );`

That would provide some more useful information to users. It would also highlight which line(s) of the source CSS errored:

![Screenshot 2023-11-15 at 10 28 25](https://github.com/WordPress/gutenberg/assets/1682452/a1e29612-4b3f-4b97-8fee-ee83a3c299ec)

zaguiini commented:
>Quickly tested and it does fix the issue.
>
>I'd recommend to add an E2E test for the feature that was actually broken in packages/block-editor/src/components/global-styles/advanced-panel.js.

Thanks. @afercia I do not have the bandwidth to add the E2E test, so the best I can do right now is to fix the unit test and merge this PR. Would you be able to add the E2E test as a follow-up PR?

zaguiini committed:
Display better warning messages

zaguiini commented:
@afercia, we have that now:

<img width="644" alt="image" src="https://github.com/WordPress/gutenberg/assets/26530524/33b794d0-9d5a-4814-815e-01daa6e2c909">

I have updated the tests and rebased the branch with `trunk`. Please re-review the PR.

zaguiini commented on packages/block-editor/src/components/global-styles/advanced-panel.js:
```diff
@@ -30,7 +30,7 @@ export default function AdvancedPanel( {
 		if ( cssError ) {
 			const [ transformed ] = transformStyles(
-				[ { css: value } ],
+				[ { css: newValue } ],
```
@afercia @youknowriad this was a bug. We were passing the existing `value` (which is composed of `{css: 'existing'}`) instead of the recently changed rules (this is the on change handler).

With this change, we're passing the correct value. If you revert it in your local environment, you'll see that `[object Object]` gets passed whenever you apply a change after blurring the field with an invalid CSS.

For the sake of expediting this change, which is breaking a couple of places, I'd say we should keep it. If we want to remove it and apply this change in a different PR, fine by me.

afercia commented:
> Would you be able to add the E2E test as a follow-up PR?

@zaguiini I'd really love to change the current red icon to a standard notice because the red icon isn't accessible at all. I will create a new issue and link to this PR as a reminder that an E2E test should be added as well.

I like the `showSourceCode` thing 🎉  I'd defer a full review to someone else though, as this isn't really my area of expertise (code looks good to me btw).

jsnajdr reviewed (approved):
Yes, this fixes the unexpected exceptions thrown by `transformStyles`. The calling code expects a `null` return value in case of error.
